compiler/rustc_hir_typeck/src/coercion.rs

```diff
@@ -1871,11 +1871,8 @@ impl<'tcx, 'exprs, E: AsCoercionSite> CoerceMany<'tcx, 'exprs, E> {
         // If this is due to a block, then maybe we forgot a `return`/`break`.
         if due_to_block
             && let Some(expr) = expression
-            && let Some((parent_fn_decl, parent_id)) = fcx
-                .tcx
-                .hir()
-                .parent_iter(block_or_return_id)
-                .find_map(|(_, node)| Some((node.fn_decl()?, node.associated_body()?.0)))
```

Comment:
Specifically this will return `None` if `node` is an item with a body, but it has no `fn_decl` (e.g. `const`s). That means that this doesn't return the parent of the body, but possibly some other ancestor, which will cause astconv to ICE.